From 204dd5db2fd949cac6ddaa9b0aa41e1480d62e1e Mon Sep 17 00:00:00 2001
From: "qipeng.zha" <qipeng.zha@intel.com>
Date: Fri, 6 Feb 2015 18:19:47 +0800
Subject: [PATCH] Touch: update handling of firmware failure

Before this change, the firmware update failure
is skipped, this will lead to kernel panic in
suspend and resume function.

Change-Id: I593d9c8ff7044410012ec5474ec2f72069a54f2c
Tracked-On: https://jira01.devtools.intel.com/browse/IMINAN-19229
Signed-off-by: qipeng.zha <qipeng.zha@intel.com>
---
 drivers/input/touchscreen/atmel_mxt_ts.c | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/drivers/input/touchscreen/atmel_mxt_ts.c b/drivers/input/touchscreen/atmel_mxt_ts.c
index c1424ef..2196bc9 100644
--- a/drivers/input/touchscreen/atmel_mxt_ts.c
+++ b/drivers/input/touchscreen/atmel_mxt_ts.c
@@ -331,7 +331,8 @@ struct mxt_data {
 #ifdef CONFIG_PM_SLEEP
 	bool power_hal_want_suspend;
 #endif
-
+	/*chip initilized status*/
+	bool initialized;
 	int alloc_pdata;
 };
 #ifdef CONFIG_PM_SLEEP
@@ -2761,6 +2762,7 @@ retry_bootloader:
 #ifdef CONFIG_DEBUG_FS
 	mxt_debugfs_create(data);
 #endif
+	data->initialized = true;
 	return 0;
 
 err_free_object_table:
@@ -3740,6 +3742,7 @@ static int mxt_probe(struct i2c_client *client,
 	data->client = client;
 	data->pdata = dev_get_platdata(&client->dev);
 	i2c_set_clientdata(client, data);
+	data->initialized = false;
 
 #ifdef CONFIG_OF
 	if (!data->pdata && client->dev.of_node)
@@ -3834,8 +3837,10 @@ static int mxt_probe(struct i2c_client *client,
 	disable_irq(data->irq);
 
 	error = mxt_initialize_with_fw_check(data);
-	if (error)
+	if (error && !data->initialized) {
+		dev_err(&client->dev, "Failure %d initialize with fw\n", error);
 		goto err_free_irq;
+	}
 
 	error = sysfs_create_group(&client->dev.kobj, &mxt_attr_group);
 	if (error) {
-- 
1.9.1

